File: NotaFiscal.Data.Mongo/NotaFiscalRepository.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Microsoft.FSharp.Core;
using MongoDB.Driver;
using MongoDB.Driver.Linq;
using static NotaFiscal.Domain.Dto;
using static Rop;

namespace NotaFiscal.Data.Mongo;

public interface INotaFiscalRepositoryErro {}

public record struct FalhaComunicarBancoDados : INotaFiscalRepositoryErro;
public record struct NotaNaoEncontrada : INotaFiscalRepositoryErro;


public interface INotaFiscalRepository
{
    Task<OperationResult<NotaFiscalServicoDto, INotaFiscalRepositoryErro>> FindByIdAsync(Guid id);
}

public class NotaFiscalRepository : INotaFiscalRepository
{
    private readonly ILogger<NotaFiscalRepository> _logger;
    private readonly IMongoCollection<NotaFiscalServicoDto> _notasFiscaisCollection;

    public NotaFiscalRepository(IOptions<MongoDbOptions> mongoDbOptions, ILogger<NotaFiscalRepository> logger)
    {
        var lol = typeof(NotaFiscalRepository);
        _logger = logger;
        var options = mongoDbOptions.Value;
        _notasFiscaisCollection = new MongoClient(options.ConnectionString)
            .GetDatabase(options.DatabaseName)
            .GetCollection<NotaFiscalServicoDto>(nameof(NotaFiscalServicoDto));
    }

    public async Task<OperationResult<NotaFiscalServicoDto, INotaFiscalRepositoryErro>> FindByIdAsync(Guid id)
    {
        try
        {
            var notaFiscal = await _notasFiscaisCollection.AsQueryable().FirstOrDefaultAsync(e => e.Id == id);
            if (notaFiscal is null)
                return fail<INotaFiscalRepositoryErro, NotaFiscalServicoDto>(new NotaNaoEncontrada());
            
            return succeed<NotaFiscalServicoDto, INotaFiscalRepositoryErro>(notaFiscal);
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Falha ao comunicar com o banco de dados");
            return fail<INotaFiscalRepositoryErro, NotaFiscalServicoDto>(new FalhaComunicarBancoDados());
        }
    }
}

public record MongoDbOptions
{
    public string ConnectionString { get; init; } = string.Empty;
    public string DatabaseName { get; init; } = string.Empty;
}

File: NotaFiscal.Data/NotaFiscalRepository.fs
namespace NotaFiscal.Data

open MongoDB.Driver
open MongoDB.Driver.Linq
open NotaFiscal.Domain.Dto.NotaFiscalServicoDto
open NotaFiscal.Domain.NotaFiscalServico
open NotaFiscal.Domain.Rop
open NotaFiscal.Domain.ApplicationErrors

module NotaFiscalRepository =
    let findNotaNotaById (collection: IMongoCollection<NotaFiscalServicoDto>) id =
        task {
            try
                let! nota = collection.AsQueryable().FirstOrDefaultAsync(fun e -> e.Id = id)
                
                return
                    match box nota with
                    | null -> NotFound (nameof NotaFiscalServico, id.ToString()) |> fail
                    | _ -> toDomain nota
                        |> mapFailuresR (FailedDeserializeFromDb >> InternalServerError)
            with ex -> return ex
                              |> (DatabaseException >> InternalServerError)
                              |> fail
        }
        
    let salvarNota (collection: IMongoCollection<NotaFiscalServicoDto>) nota =
        task {
            try
                do! fromDomain nota |> collection.InsertOneAsync
                return succeed ()
            with ex -> return ex
                              |> (DatabaseException >> InternalServerError)
                              |> fail 
        }